src/vm/sampleprofiler.cpp

```diff
@@ -103,34 +120,35 @@ void SampleProfiler::Disable()
         GC_TRIGGERS;
         MODE_ANY;
         // Synchronization of multiple callers occurs in EventPipe::Disable.
-        PRECONDITION(EventPipe::GetLock()->OwnedByCurrentThread());
+        PRECONDITION(EventPipe::IsLockOwnedByCurrentThread());
+        PRECONDITION(s_RefCount >= 0);
     }
     CONTRACTL_END;
 
     // Bail early if profiling is not enabled.
-    if(!s_profilingEnabled)
-    {
+    if (!s_profilingEnabled)
         return;
-    }
 
-    _ASSERTE(!g_fProcessDetach);
-    
-    // Reset the event before shutdown.
-    s_threadShutdownEvent.Reset();
+    if (s_RefCount == 1)
+    {
+        _ASSERTE(!g_fProcessDetach);
 
-    // The sampling thread will watch this value and exit
-    // when profiling is disabled.
-    s_profilingEnabled = false;
+        // Reset the event before shutdown.
+        s_threadShutdownEvent.Reset();
```

Comment:
I would reset the event prior to creating to the sampling thread. Otherwise any error that causes the thread to exit prematurely will set the event, then it will be reset here, and the Wait below will block forever.